### PR TITLE
rootless: Support BuildKit containerd worker

### PR DESCRIPTION
**Pull request by ktock**
Fixes: #434

This commit patches `containerd-rootless-setuptool.sh` to allow configuring nerdctl for BuildKit rootless containerd worker (BuildKit v0.10).

This enables nerdctl to share images into BuildKit which isn't possible with OCI worker (#434).
For example, this enables BuildKit to use images pulled by `nerdctl pull` as base images.

```console
$ nerdctl pull ghcr.io/stargz-containers/ubuntu:20.04-org
$ nerdctl tag ghcr.io/stargz-containers/ubuntu:20.04-org mybase
$ mkdir -p /tmp/ctx && cat <<EOF > /tmp/ctx/Dockerfile
FROM mybase
RUN echo hello > /hello
CMD ["cat", "/hello"]
EOF
$ nerdctl build -t mytest /tmp/ctx
[+] Building 5.4s (6/6) FINISHED                                                                                                                                                                                                                                                         
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                0.1s
 => => transferring dockerfile: 95B                                                                                                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                   0.2s
 => => transferring context: 2B                                                                                                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/mybase:latest                                                                                                                                                                                                                    2.6s
 => [1/2] FROM docker.io/library/mybase@sha256:adf73ca014822ad8237623d388cedf4d5346aa72c270c5acc01431cc93e18e2d                                                                                                                                                                     1.5s
 => => resolve docker.io/library/mybase@sha256:adf73ca014822ad8237623d388cedf4d5346aa72c270c5acc01431cc93e18e2d                                                                                                                                                                     1.3s
 => [2/2] RUN echo hello > /hello                                                                                                                                                                                                                                                   0.2s
 => exporting to oci image format                                                                                                                                                                                                                                                   0.8s
 => => exporting layers                                                                                                                                                                                                                                                             0.2s
 => => exporting manifest sha256:4b8d0218edd5c342e1a7c3bc9cb3adfeec480563ff06535b85c0fd85fc1bbad2                                                                                                                                                                                   0.0s
 => => exporting config sha256:b6cbc7ea26b1fd045d589f7fef3d011165f3b0f146c32ca6bae06d47199a5a7b                                                                                                                                                                                     0.0s
 => => sending tarball                                                                                                                                                                                                                                                              0.5s
unpacking docker.io/library/mytest:latest (sha256:4b8d0218edd5c342e1a7c3bc9cb3adfeec480563ff06535b85c0fd85fc1bbad2)...done
$ nerdctl run --rm -it mytest
hello
```

Note that 1 containerd worker can use 1 containerd namespace so buildkitd needs to be launched for each namespace the user wants to use.

## setup (rootless)

Rootless user can enable containerd worker (in "default" namespace of containerd) like the following:

```console
$ CONTAINERD_NAMESPACE=default containerd-rootless-setuptool.sh install-buildkit-containerd
```

NOTE1: `containerd-rootless-setuptool.sh` is aware of `CONTAINERD_NAMESPACE` and `CONTAINERD_SNPASHOTTER`.
User can launch additional buildkitd processes by executing this script with different `CONTAINERD_NAMESPACE` values.
NOTE2: If `CONTAINERD_NAMESPACE` is not specified, buildkitd installed in "buildkit" namespace which is the default of buildkitd. If a user wants to give buildkitd access to the images in "default" namespace, this command should be run with `CONTAINERD_NAMESPACE=default`.

When $BUILDKIT_HOST is unset, nerdctl (both rootless and rootful) tries to use `buildkit-<NS>/buildkitd.sock` by default and then try to fall back to `buildkit-default/buildkitd.sock`, `buildkit/buildkitd.sock`

~Maybe we can add an option to allow rootless nerdctl automatically detecting BuildKit socket location for a namespace (i.e. `unix://${XDG_RUNTIME_DIR}/buildkit-${CONTAINERD_NAMESPACE}/buildkitd.sock`) without manually configuring `BUILDKIT_HOST`.~ **Added this**

## setup (rootful)

Rootful containerd worker has been available even without this patch.
Use the configuration (`/etc/buildkit/buildkitd.toml`) suggested at https://github.com/containerd/nerdctl/issues/434#issuecomment-998702312.

```toml
[worker.oci]
  enabled = false

[worker.containerd]
  enabled = true
  # namespace should be "k8s.io" for Kubernetes (including Rancher Desktop)
  namespace = "default"
```

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -84,8 +84,10 @@ jobs:
     strategy:
       matrix:
         containerd: [v1.4.5, v1.6.1, main]
+        buildkit-worker: [oci, containerd]
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:87]: We seriously have to deflake CI before expanding the test matrix

https://github.com/containerd/nerdctl/issues?q=is%3Aissue+is%3Aopen+label%3Aarea%2Fci

Comment (AkihiroSuda) [reply .github/workflows/test.yml:87]: Until that we can just test the containerd worker (after BK v0.10.0 GA)

File: cmd/nerdctl/build_test.go [hunk 2]
```diff
@@ -48,6 +48,104 @@ CMD ["echo", "nerdctl-build-test-string"]
...
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx2)
+
+	base.Cmd("build", "-t", imageName2, buildCtx2).AssertOK()
+	base.Cmd("build", buildCtx2, "-t", imageName2).AssertOK()
+
+	base.Cmd("run", "--rm", imageName2).AssertOutExactly("hello2\n")
+}
+
+func TestBuildUnlazy(t *testing.T) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build_test.go:125]: What is "Unlazy" here? Doesn't seem related to stargz stuff though

File: cmd/nerdctl/build_test.go [hunk 3]
```diff
@@ -48,6 +48,104 @@ CMD ["echo", "nerdctl-build-test-string"]
 	base.Cmd("run", "--rm", imageName).AssertOutExactly("nerdctl-build-test-string\n")
 }
 
+func TestBuildBaseImage(t *testing.T) {
+	if os.Getenv("TEST_BUILDKIT_WORKER") != "containerd" {
+		t.Skip()
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build_test.go:53]: Needs explanation

File: extras/rootless/containerd-rootless-setuptool.sh [hunk 4]
```diff
@@ -273,6 +273,51 @@ cmd_entrypoint_install_buildkit() {
...
+	ADDRESS="${XDG_RUNTIME_DIR}/buildkit-${NAMESPACE}/buildkitd.sock"
+	ROOT="${XDG_DATA_HOME}/buildkit-${NAMESPACE}"
+	cat <<-EOT | install_systemd_unit "${NAMESPACE}-${SYSTEMD_BUILDKIT_UNIT}"
+		[Unit]
+		Description=BuildKit (Rootless)
+		PartOf=${SYSTEMD_CONTAINERD_UNIT}
+
+		[Service]
+		Environment=PATH=$BIN:/sbin:/usr/sbin:$PATH
+		ExecStart="$REALPATH0" nsenter -- buildkitd --config=${XDG_CONFIG_HOME}/buildkit/buildkitd.toml --addr="unix://${ADDRESS}" --root="${ROOT}" --containerd-worker-namespace="${NAMESPACE}" --containerd-worker-snapshotter="${CONTAINERD_SNAPSHOTTER:-overlayfs}"
```
Comment (AkihiroSuda) [inline extras/rootless/containerd-rootless-setuptool.sh:308]: No need to specify the toml AFAICS

File: extras/rootless/containerd-rootless-setuptool.sh [hunk 5]
```diff
@@ -273,6 +273,51 @@ cmd_entrypoint_install_buildkit() {
...
+		ExecReload=/bin/kill -s HUP \$MAINPID
+		RestartSec=2
+		Restart=always
+		Type=simple
+		KillMode=mixed
+
+		[Install]
+		WantedBy=default.target
+	EOT
+	INFO "Set \`export BUILDKIT_HOST=\"unix://${ADDRESS}\"\` to use this configuration of buildkitd"
```
Comment (AkihiroSuda) [inline extras/rootless/containerd-rootless-setuptool.sh:318]: When `$BUILDKIT_HOST` is unset, nerdctl (both rootless and rootful) should try to use `buildkit-<NS>/buildkitd.sock` by default and then try to fall back to `buildkit-default/buildkitd.sock`, `buildkit/buildkitd.sock`

Comment (AkihiroSuda) [reply extras/rootless/containerd-rootless-setuptool.sh:318]: Also, `nerdctl build` should try to use the containerd `image` exporter when the worker UUID corresponds to the containerd UUID && the worker namespace corresponds to the current namespace

File: .github/workflows/test.yml (hunk 1 above)
Comment (ktock) [reply .github/workflows/test.yml]: Enabled the test only with containerd worker

File: cmd/nerdctl/build_test.go (hunk 2 above)
Comment (ktock) [reply cmd/nerdctl/build_test.go]: Deleted this test because TestBuild already coverts this.

What I wanted to test was the following edge case:

If a build doesn't run any commands nor perform COPY on the base image, BuildKit tries not to download that image (i.e. keeps it lazy). This behaviour happens not only with stargz-snapshotter but also with any type of snapshotter.
`type=oci`/`type=docker` triggers unlazying of that base image but `type=image` doesn't. This leads to "not found" error for the contents of that image e.g. when nerdctl tries to run it. As a workaround, nerdctl adds `unpack=true` to `buildctl build` which triggers unlazy even when `type=image` is specified.

TestBuild is exactly the situation where the base image can be lazy. So this test is enough for testing that this workaround actually works.

File: cmd/nerdctl/build_test.go (hunk 3 above)
Comment (ktock) [reply cmd/nerdctl/build_test.go]: Now we don't need this envvar because we always run tests with containerd worker.

File: extras/rootless/containerd-rootless-setuptool.sh (hunk 4 above)
Comment (ktock) [reply extras/rootless/containerd-rootless-setuptool.sh]: Fixed this.

File: extras/rootless/containerd-rootless-setuptool.sh (hunk 5 above)
Comment (ktock) [reply extras/rootless/containerd-rootless-setuptool.sh]: Added this feature.

File: extras/rootless/containerd-rootless-setuptool.sh [hunk 6]
```diff
@@ -273,6 +273,50 @@ cmd_entrypoint_install_buildkit() {
...
+	ADDRESS="${XDG_RUNTIME_DIR}/buildkit-${NAMESPACE}/buildkitd.sock"
+	ROOT="${XDG_DATA_HOME}/buildkit-${NAMESPACE}"
+	cat <<-EOT | install_systemd_unit "${NAMESPACE}-${SYSTEMD_BUILDKIT_UNIT}"
+		[Unit]
+		Description=BuildKit (Rootless)
+		PartOf=${SYSTEMD_CONTAINERD_UNIT}
+
+		[Service]
+		Environment=PATH=$BIN:/sbin:/usr/sbin:$PATH
+		ExecStart="$REALPATH0" nsenter -- buildkitd --addr="unix://${ADDRESS}" --root="${ROOT}" --containerd-worker-namespace="${NAMESPACE}" --containerd-worker-snapshotter="${CONTAINERD_SNAPSHOTTER:-overlayfs}"
```
Comment (AkihiroSuda) [inline extras/rootless/containerd-rootless-setuptool.sh:308]: Just leave --containerd-worker-snapshotter unset when $CONTAINERD_SNAPSHOTTER is not set

Comment (AkihiroSuda) [reply extras/rootless/containerd-rootless-setuptool.sh:308]: Same applies to other flags too

File: cmd/nerdctl/system.go [hunk 7]
```diff
@@ -33,6 +39,40 @@ func newSystemCommand() *cobra.Command {
 	systemCommand.AddCommand(
 		newEventsCommand(),
 		newInfoCommand(),
+		newBuildPruneCommand(),
```
Comment (AkihiroSuda) [inline cmd/nerdctl/system.go:42]: This has to be `nerdctl builder prune`
https://docs.docker.com/engine/reference/commandline/builder_prune/

File: extras/rootless/containerd-rootless-setuptool.sh [hunk 8]
```diff
@@ -273,6 +273,58 @@ cmd_entrypoint_install_buildkit() {
...
+		ERROR "Install containerd first (\`$ARG0 install\`)"
+		exit 1
+	fi
+	UNIT_NAME=${SYSTEMD_BUILDKIT_UNIT}
+	BUILDKITD_FLAG=
+	if [ -n "${CONTAINERD_NAMESPACE:-}" ] ; then
+		UNIT_NAME="${CONTAINERD_NAMESPACE}-${SYSTEMD_BUILDKIT_UNIT}"
+		BUILDKITD_FLAG="${BUILDKITD_FLAG} --addr=unix://${XDG_RUNTIME_DIR}/buildkit-${CONTAINERD_NAMESPACE}/buildkitd.sock --root=${XDG_DATA_HOME}/buildkit-${CONTAINERD_NAMESPACE} --containerd-worker-namespace=${CONTAINERD_NAMESPACE}"
+	else
+		WARNING "buildkitd has access images in \"buildkit\" namespace by default. If you want to give buildkitd access to the images in \"default\" namespace, run this command with CONTAINERD_NAMESPACE=default"
```
Comment (AkihiroSuda) [inline extras/rootless/containerd-rootless-setuptool.sh:304]:
```suggestion
		WARNING "buildkitd has access to images in \"buildkit\" namespace by default. If you want to give buildkitd access to the images in \"default\" namespace, run this command with CONTAINERD_NAMESPACE=default"
```

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
We need to have docs about this later

**Comment by AkihiroSuda** [comment]
Regression: OCI worker (`/run/user/1001/buildkit/buildkitd.sock`) can no longer be auto detected:

```console
$ nerdctl build -t foo .
ERRO[0000] error: failed to list workers: Unavailable: connection error: desc = "transport: error while dialing: dial unix /run/user/1001/buildkit-default/buildkitd.sock: connect: no such file or directory" 
WARN[0000] `buildctl` needs to be installed and `buildkitd` needs to be running, see https://github.com/moby/buildkit , and `containerd-rootless-setuptool.sh install-buildkit` for OCI worker or `containerd-rootless-setuptool.sh install-buildkit-containerd` for containerd worker: exit status 1  host="unix:///run/user/1001/buildkit-default/buildkitd.sock"
```

File: cmd/nerdctl/build.go [hunk 9]
```diff
@@ -73,22 +74,59 @@ func newBuildCommand() *cobra.Command {
...
+	}
+	containerdUUID, ok := labels["org.mobyproject.buildkit.worker.containerd.uuid"]
+	if !ok {
+		return false, nil
+	}
+	containerdNamespace, ok := labels["org.mobyproject.buildkit.worker.containerd.namespace"]
+	if !ok {
+		return false, nil
+	}
+	return executor == "containerd" && containerdUUID == uuid && containerdNamespace == namespace, nil
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:114]: Maybe we should compare the snapshotter too?
To avoid some potential issues with unpacked images.

Comment (AkihiroSuda) [reply cmd/nerdctl/build.go:114]: cc @ktock

Comment (ktock) [reply cmd/nerdctl/build.go:114]: SGTM
